// redux-form/FieldListItem/index.js
import React from 'react'
import PropTypes from 'prop-types'

/**
 * Компонент элемента списка
 *
 * Принимает список дочерних полей,
 * передавая им необходимые аргументы
 *
 * @param {Number}        id
 * @param {String}        member        Название элемента списка. Из redux-form
 * @param {Function}      onRemove      Функция, вызываемая при удалении
 * @param {React.Element} buttonElement Элемент кнопки "Удалить"
 */
const FieldListItem = ({
  id,
  member,
  onRemove,
  buttonElement = <button>X</button>,
  children,
  ...props
}) => {
  const childArray = React.Children.toArray(children)
  return (
    <div {...props}>
      {
        childArray.map(
          child => React.cloneElement(
            child,
            { name: `${member}.${child.props.name}` }
          )
        )
      }
      {
        React.cloneElement(
          buttonElement,
          {
            onClick: evt => {
              evt.preventDefault()
              onRemove(id)
            }
          }
        )
      }
    </div>
  )
}

FieldListItem.propTypes = {
  id: PropTypes.number,
  member: PropTypes.string.isRequired,
  onRemove: PropTypes.func,
  buttonElement: PropTypes.element,
  children: PropTypes.node.isRequired
}

export default FieldListItem
